src/Posts.tsx

```diff
@@ -1,50 +1,214 @@
   modified: Date;
 };
+type Category = {
+  id: number;
+  name: string;
+  description: string;
+};
 
 class Posts extends React.Component<Props, State> {
   constructor(props: Props) {
     super(props);
     this.state = {
       posts: [],
       isLoading: false,
-      errors: []
+      errors: undefined,
+      newPostErrors: '',
+      categories: [],
+      formIsValid: false,
+      topicValue: '',
+      topicValid: false,
+      postValue: '',
+      postValid: false,
+      pinnedValue: false,
+      categoryValue: '1',
+      postsLoading: false,
+      categoriesLoading: false,
+      topicError: '',
+      postError: ''
     };
   }
   componentDidMount() {
-    this.getAndSetPosts();
+    this.loadPostsAndCategories();
   }
+  loadPostsAndCategories = async () => {
+    this.setState({ isLoading: true, errors: undefined });
+    await this.getAndSetPosts();
+    await this.getCategories();
+  };
   getAndSetPosts = async () => {
-    this.setState({ isLoading: true, errors: [] });
-    fetch(API)
+    this.setState({ postsLoading: true });
+    fetch(API + 'posts')
+      .then(response => {
+        if (response.status === 404) {
+          return [];
+        } else if (response.ok) {
+          return response.json();
+        } else {
+          throw Error(response.statusText);
+        }
+      })
+      .then(data => this.setState({ posts: data, postsLoading: false }))
+      .catch(errors => this.setState({ errors, postsLoading: false }))
+      .then(() => this.isBothLoaded());
+  };
+  getCategories = async () => {
+    this.setState({ categoriesLoading: true });
+    fetch(API + 'categories')
+      .then(response => {
+        if (response.status === 404) {
+          return [];
+        } else if (response.ok) {
+          return response.json();
+        } else {
+          throw Error(response.statusText);
+        }
+      })
+      .then(data => this.setState({ categories: data, categoriesLoading: false }))
+      .catch(errors => this.setState({ errors, categoriesLoading: false }))
+      .then(() => this.isBothLoaded());
+  };
+  saveNewPost = async () => {
+    const data = {
+      post: this.state.postValue,
+      topic: this.state.topicValue,
+      pinned: this.state.pinnedValue,
+      categoryId: parseInt(this.state.categoryValue)
+    };
+
+    fetch(API + 'posts', {
+      method: 'POST',
+      mode: 'cors',
+      cache: 'no-cache',
+      credentials: 'same-origin',
+      headers: {
+        'Content-Type': 'application/json'
+      },
+      referrer: 'no-referrer',
+      body: JSON.stringify(data)
+    })
       .then(response => {
         if (!response.ok) {
           throw Error(response.statusText);
         } else {
           return response.json();
         }
       })
-      .then(data => this.setState({ posts: data, isLoading: false }))
-      .catch(errors => this.setState({ errors, isLoading: false }));
+      .then(data => this.setState({ posts: [...this.state.posts, data] }))
```

Comment:
Same kind of treatment here that maybe add new post to the first place of the new list